doc/usbguard.1
.\" Automatically generated by Pandoc 1.19.1
.\"
.TH "USBGUARD" "1" "April 2017" "" ""
.hy
.SH NAME
.PP
\f[B]usbguard\f[] \-\- USBGuard command\-line interface
.SH SYNOPSIS
.PP
usbguard\ [\f[I]OPTIONS\f[]]\ \f[I]command\f[]\ [\f[I]COMMAND\-OPTIONS\f[]]\ ...
.PP
usbguard\ \f[B]list\-devices\f[]
.PP
usbguard\ \f[B]allow\-device\f[]\ <\f[I]id\f[]>
.PP
usbguard\ \f[B]block\-device\f[]\ <\f[I]id\f[]>
.PP
usbguard\ \f[B]reject\-device\f[]\ <\f[I]id\f[]>
.PP
usbguard\ \f[B]list\-rules\f[]
.PP
usbguard\ \f[B]append\-rule\f[]\ <\f[I]rule\f[]>
.PP
usbguard\ \f[B]remove\-rule\f[]\ <\f[I]id\f[]>
.PP
usbguard\ \f[B]generate\-policy\f[]
.PP
usbguard\ \f[B]watch\f[]
.PP
usbguard \f[B]read\-descriptor\f[] <\f[I]file\f[]>
.PP
usbguard \f[B]add\-user\f[] <\f[I]name\f[]>
.PP
usbguard \f[B]remove\-user\f[] <\f[I]name\f[]>
.SH DESCRIPTION
.PP
The \f[B]usbguard\f[] command provides a command\-line interface (CLI)
to the \f[B]usbguard\-daemon\f[](8) instance and provides a tool for
generating initial USBGuard policies.
.SH SUBCOMMANDS
.PP
\f[B]list\-devices\f[]
.PP
List all USB devices recognized by the USBGuard daemon.
.PP
Available options:
.TP
.B \f[B]\-a\f[], \f[B]\-\-allowed\f[]
List allowed devices.
.RS
.RE
.TP
.B \f[B]\-b\f[], \f[B]\-\-blocked\f[]
List blocked devices.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]allow\-device\f[] [\f[I]OPTIONS\f[]] <\f[I]id\f[]>
.PP
Authorize a device identified by the device \f[I]id\f[] to interact with
the system.
.PP
Available options:
.TP
.B \f[B]\-p\f[], \f[B]\-\-permanent\f[]
Make the decision permanent.
A device specific allow rule will be appended to the current policy.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]block\-device\f[] [\f[I]OPTIONS\f[]] <\f[I]id\f[]>
.PP
Deauthorize a device identified by the device \f[I]id\f[].
.PP
Available options:
.TP
.B \f[B]\-p\f[], \f[B]\-\-permanent\f[]
Make the decision permanent.
A device specific block rule will be appended to the current policy.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]reject\-device\f[] [\f[I]OPTIONS\f[]] <\f[I]id\f[]>
.PP
Deauthorize and remove a device identified by the device \f[I]id\f[].
.PP
Available options:
.TP
.B \f[B]\-p\f[], \f[B]\-\-permanent\f[]
Make the decision permanent.
A device specific reject rule will be appended to the current policy.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]list\-rules\f[] [\f[I]OPTIONS\f[]]
.PP
List the rule set (policy) used by the USBGuard daemon.
.PP
Available options:
.TP
.B \f[B]\-d\f[], \f[B]\-\-show\-devices\f[]
Show all devices which are affected by the specific rule.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]append\-rule\f[] [\f[I]OPTIONS\f[]] <\f[I]rule\f[]>
.PP
Append the \f[I]rule\f[] to the current rule set.
.PP
Available options:
.TP
.B \f[B]\-a\f[], \f[B]\-\-after\f[] <\f[I]id\f[]>
Append the new rule after a rule with the specified rule \f[I]id\f[].
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]remove\-rule\f[] [\f[I]OPTIONS\f[]] <\f[I]id\f[]>
.PP
Remove a rule identified by the rule \f[I]id\f[] from the rule set.
.PP
Available options:
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]generate\-policy\f[] [\f[I]OPTIONS\f[]]
.PP
Generate a rule set (policy) which authorizes the currently connected
USB devices.
.PP
Available options:
.TP
.B \f[B]\-p\f[], \f[B]\-\-with\-ports\f[]
Generate port specific rules for all devices.
By default, port specific rules are generated only for devices which do
not export an iSerial value.
.RS
.RE
.TP
.B \f[B]\-P\f[], \f[B]\-\-no\-ports\-sn\f[]
Don\[aq]t generate port specific rules for devices without an iSerial
value.
Without this option, the tool will add a via\-port attribute to any
device that doesn\[aq]t provide a serial number.
This is a security measure to limit devices that cannot be uniquely
identified to connect only via a specific port.
This makes it harder to bypass the policy since the real device will
occupy the allowed USB port most of the time.
.RS
.RE
.TP
.B \f[B]\-t\f[], \f[B]\-\-target\f[] <\f[I]target\f[]>
Generate an explicit "catch all" rule with the specified target.
The target can be one of the following values: \f[B]allow\f[],
\f[B]block\f[], \f[B]reject\f[]
.RS
.RE
.TP
.B \f[B]\-X\f[], \f[B]\-\-no\-hashes\f[]
Don\[aq]t generate a hash attribute for each device.
.RS
.RE
.TP
.B \f[B]\-H\f[], \f[B]\-\-hash\-only\f[]
Generate a hash\-only policy.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]watch\f[] [\f[I]OPTIONS\f[]]
.PP
Watch the IPC interface events and print them to stdout.
.PP
Available options:
.TP
.B \f[B]\-w\f[], \f[B]\-\-wait\f[]
Wait for IPC connection to become available.
.RS
.RE
.TP
.B \f[B]\-o\f[], \f[B]\-\-once\f[]
Wait only when starting, if needed.
Exit when the connection is lost.
.RS
.RE
.TP
.B \f[B]\-e\f[], \f[B]\-\-exec\f[] <\f[I]path\f[]>
Run an executable file located at \f[I]path\f[] for every event.
Pass event data to the process via environment variables.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]read\-descriptor\f[] [\f[I]OPTIONS\f[]] <\f[I]file\f[]>
.PP
Read a USB descriptor from a file and print it in human\-readable form.
.PP
Available options:
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
~ ~ ~
.RS
.RE
.PP
\f[B]add\-user\f[] <\f[I]name\f[]> [\f[I]OPTIONS\f[]]
.PP
Create an IPC access control file allowing the user/group identified by
\f[I]name\f[] to use the USBGuard IPC bus.
The change takes effect only after restarting the
\f[B]usbguard\-daemon\f[](8) instance.
.PP
Available options:
.TP
.B \f[B]\-u\f[], \f[B]\-\-user\f[]
The specified \f[I]name\f[] represents a username or UID (default).
.RS
.RE
.TP
.B \f[B]\-g\f[], \f[B]\-\-group\f[]
The specified \f[I]name\f[] represents a groupname or GID.
.RS
.RE
.TP
.B \f[B]\-p\f[], \f[B]\-\-policy\f[] <\f[I]privileges\f[]>
Policy related privileges.
.RS
.RE
.TP
.B \f[B]\-d\f[], \f[B]\-\-devices\f[] <\f[I]privileges\f[]>
Device related privileges.
.RS
.RE
.TP
.B \f[B]\-e\f[], \f[B]\-\-exceptions\f[] <\f[I]privileges\f[]>
Exceptions related privileges.
.RS
.RE
.TP
.B \f[B]\-P\f[], \f[B]\-\-parameters\f[] <\f[I]privileges\f[]>
Run\-time parameter related privileges.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
.PP
Privileges:
.PP
When specifying \f[I]privileges\f[], they are expected to be in the form
of a list separated by a colon:
.IP
.nf
\f[C]
\ \ \ \ $\ sudo\ usbguard\ add\-user\ joe\ \-\-devices=listen,modify
\f[]
.fi
.TP
.B Consult the \f[B]usbguard\-daemon.conf\f[](5) man\-page for a detailed list of available privileges in each section.
~ ~ ~
.RS
.RE
.PP
\f[B]remove\-user\f[] <\f[I]name\f[]> [\f[I]OPTIONS\f[]]
.PP
Remove an IPC access control file associated with the user/group
identified by \f[I]name\f[].
The change takes effect only after restarting the
\f[B]usbguard\-daemon\f[](8) instance.
.PP
Available options:
.TP
.B \f[B]\-u\f[], \f[B]\-\-user\f[]
The specified \f[I]name\f[] represents a username or UID (default).
.RS
.RE
.TP
.B \f[B]\-g\f[], \f[B]\-\-group\f[]
The specified \f[I]name\f[] represents a groupname or GID.
.RS
.RE
.TP
.B \f[B]\-h\f[], \f[B]\-\-help\f[]
Show help.
.RS
.RE
.SH EXAMPLES
.PP
\f[B]Creating an initial policy\f[]
.IP
.nf
\f[C]
\ \ \ \ $\ sudo\ usbguard\ generate\-policy\ >\ rules.conf
\ \ \ \ $\ vi\ rules.conf
\ \ \ \ (review/modify\ the\ rule\ set)
\ \ \ \ $\ sudo\ install\ \-m\ 0600\ \-o\ root\ \-g\ root\ \\
\ \ \ \ \ \ \ rules.conf\ /etc/usbguard/rules.conf
\f[]
.fi
.SH BUGS
.PP
If you find a bug in this software or if you\[aq]d like to request a
feature to be implemented, please file a ticket at
<https://github.com/dkopecek/usbguard/issues/new>.
.SH COPYRIGHT
.PP
Copyright © 2017 Red Hat, Inc.
License GPLv2+: GNU GPL version 2 or later
<http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH SEE ALSO
.PP
\f[B]usbguard\-rules.conf\f[](5), \f[B]usbguard\-daemon\f[](8),
\f[B]usbguard\-daemon.conf\f[](5)
.SH AUTHORS
Daniel Kopeček <dkopecek@redhat.com>.


CHANGELOG.md
# Change Log

## 0.7.1 - <unrelased>

### Added
- CLI: usbguard watch command now includes an -e <path> option to run an executable
  for every received event. Event data are passed to the executable via environment
  variables.
- usbguard-daemon: added "-K" option which can disable logging to console.
- Added zsh autocompletion support

### Removed

### Changed
- Qt Applet: disabled session management
- usbguard-daemon console logging output is enabled by default now.   Previously,
  the -k option had to be passed to enable the output.
- Replaced --enable-maintainer-mode configure option with --enable-full-test-suite
  option. When the new option is not used during the configure phase, only a basic
  set of test is run during the make check phase.

## 0.7.0 - 2017-04-12
### Added
- Added InsertedDevicePolicy configuration option to control the policy
  method for inserted devices.
- Added RestoreControllerDeviceState configuration option.
- Added DeviceManagerBackend configuration option. This option can be used
  to select from several device manager backend implementations.
- Implemented an uevent based device manager backend.
- Added setParameter, getParameter IPC (incl. D-Bus) methods.
- Added set-parameter, get-parameter CLI subcommands.
- Qt Applet: Added Spanish (es_AR) translation.
- Create empty rules.conf file at install time (make install).
- Support for numeric UID/GID values in IPCAllowedUsers and IPCAllowedGroups
  settings.
- If bash completion support is detected at configure time, install the bash
  completion script during make install.
- Added new configuration setting: IPCAccessControlFiles.
- IPC access is now configurable down to a section and privilege
  level per user and/or group.
- Added add-user, remove-user usbuard CLI subcommands for creating, removing
  IPC access control files.
- Added AuditFilePath configuration option for setting the location of the
  USBGuard audit events log file path. If set, the usbguard-daemon will log
  policy and device related actions and whether they succeeded or not.

### Removed
- Removed UDev based device manager backend and UDev related dependencies.
- Removed UDev development files/API dependecy

### Changed
- Reset Linux root hub bcdDevice value before updating device hash. This is
  a backwards incompatible change because it changes how the device hash is
  computed for Linux root hub devices.
- Refactored low-level USB device handling into SysFSDevice class which
  represents a device in the /sys filesystem (sysfs).
- Removed usage of `readdir_r` because it's obsolete. Replaced with readdir
  with the assumption that its usage is thread-safe if the directory handle
  passed to it is not shared between threads.
- Extended test suite with use case tests.
- Install the usbguard-daemon configuration and policy file with strict file
  permissions to prevent policy leaks.
- Fixed several memory leaks.
- Don't pre-resolve user and group names in IPCAllowedUsers and IPCAllowedGroups
  settings. Instead, resolve the name during the IPC authentication phase.
- Changed default values of PresentDevicePolicy (to apply-policy) and
  IPCAllowedGroups (removed wheel group entry) settings in the shipped
  usbguard-daemon.conf configuration file.

## 0.6.2 - 2016-09-18
### Changed
- Wait for disconnect in IPCClient dtor if needed
- Qt Applet: Fixed loading of decision method and default decision settings

## 0.6.1 - 2016-09-16
### Changed
- Refactored logging subsystem
- Fixed handling of IPC disconnect in the IPCClient class
- Qt Applet: Fixed handling of main window minimization and maximization
- Fixed building on architectures that don't provide required atomic operations.
  The libatomic emulation library will be used in such cases.
- Fixed several typos in the documentation

### Added
- Implemented a simple internal logger
- Access to the logger via public API
- Improved logging coverage. Logging output can be enabled either via
  CLI options or by setting the `USBGUARD_DEBUG` environment variable to 1.
- Qt Applet: UI translation support.
- Qt Applet: Czech (cs_CZ) translation

### Removed
- Removed spdlog dependency

## 0.6.0 - 2016-09-04
### Changed
- Fixed the daemon to use the match target when matching device rules.
  This fixes matching devices against the rules in the policy.
- Switched to protobuf based IPC
- Simplified the IPC and D-Bus interfaces
- Refactored custom exceptions

### Added
- Added DevicePresenceChanged signal for notifying about device
  insertions, updates and removals
- Added DevicePolicyChanged signal for notifying about device
  policy changes.
- Added ExceptionMessage signal for sending asynchronous exceptions
  to the IPC/D-Bus clients.
- Extended the usbguard CLI watch subcommand with the ability to
  wait for the IPC connection to become available. See the new `-w`
  and `-o` options for details.

### Removed
- Removed nlohmann/json submodule
- Removed allowDevice, blockDevice and rejectDevice methods from
  the IPC and D-Bus interface.
- Removed DeviceAllowed, DeviceBlocked and DeviceRejected signals
  from the IPC and D-Bus interface (replaced by single signal)
- Removed DeviceInserted, DevicePresent and DeviceRemoved signals
  from the IPC and D-Bus interface (replaced by single signal)

## 0.5.14 - 2016-08-18
### Changed
- Fixed unknown descriptor type handling

## 0.5.13 - 2016-08-16
### Changed
- refactored USB parser to support multiple handlers per USB descriptor type
- Qt Applet: device list should be always expanded

### Added
- added support for an audio device related endpoint descriptor

## 0.5.12 - 2016-08-14
### Changed
- Fixed a bug in matching USB interface types with wildcards
- usbguard-daemon will now abort at startup if the rule file contains
  syntax error instead of continuing with an empty rule set

## 0.5.11 - 2016-08-13
### Added
- Maintainer script for spellchecking source archive files
- Rule::Attribute class for representing the rule attributes

src/CLI/usbguard.cpp
//
// Copyright (C) 2016 Red Hat, Inc.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Authors: Daniel Kopecek <dkopecek@redhat.com>
//
#ifdef HAVE_BUILD_CONFIG_H
#include <build-config.h>
#endif

#include "usbguard/Logger.hpp"
#include "usbguard/Exception.hpp"

#include <map>
#include <iostream>

#ifndef _GNU_SOURCE
# define _GNU_SOURCE
#endif
#include <cstring> /* GNU version of basename(3) */

#include "usbguard.hpp"
#include "usbguard-get-parameter.hpp"
#include "usbguard-set-parameter.hpp"
#include "usbguard-list-devices.hpp"
#include "usbguard-list-rules.hpp"
#include "usbguard-generate-policy.hpp"
#include "usbguard-allow-device.hpp"
#include "usbguard-block-device.hpp"
#include "usbguard-reject-device.hpp"
#include "usbguard-append-rule.hpp"
#include "usbguard-remove-rule.hpp"
#include "usbguard-watch.hpp"
#include "usbguard-read-descriptor.hpp"
#include "usbguard-add-user.hpp"
#include "usbguard-remove-user.hpp"

namespace usbguard
{
  const char *usbguard_arg0 = nullptr;

  static const std::map<const std::string,int(*)(int, char**)> cmd_handler_map = {
    { "get-parameter", &usbguard_get_parameter },
    { "set-parameter", &usbguard_set_parameter },
    { "list-devices", &usbguard_list_devices },
    { "allow-device", &usbguard_allow_device },
    { "block-device", &usbguard_block_device },
    { "reject-device", &usbguard_reject_device },
    { "list-rules", &usbguard_list_rules },
    { "append-rule", &usbguard_append_rule },
    { "remove-rule", &usbguard_remove_rule },
    { "generate-policy", &usbguard_generate_policy },
    { "watch", &usbguard_watch },
    { "read-descriptor", &usbguard_read_descriptor },
    { "add-user", &usbguard_add_user },
    { "remove-user", &usbguard_remove_user }
  };

  static void showTopLevelHelp(std::ostream& stream = std::cout)
  {
    stream << " Usage: " << usbguard_arg0
           << " [OPTIONS] <command> [COMMAND OPTIONS] ..." << std::endl;
    stream << std::endl;
    stream << " Options:" << std::endl;
    stream << "" << std::endl;
    stream << " Commands:" << std::endl;
    stream << "  get-parameter <name>           Get the value of a runtime parameter." << std::endl;
    stream << "  set-parameter <name> <value>   Set the value of a runtime parameter." << std::endl;
    stream << "  list-devices                   List all USB devices recognized by the USBGuard daemon." << std::endl;
    stream << "  allow-device <id>              Authorize a device to interact with the system." << std::endl;
    stream << "  block-device <id>              Deauthorize a device." << std::endl;
    stream << "  reject-device <id>             Deauthorize and remove a device from the system." << std::endl;
    stream << std::endl;
    stream << "  list-rules                     List the rule set (policy) used by the USBGuard daemon." << std::endl;
    stream << "  append-rule <rule>             Append a rule to the rule set." << std::endl;
    stream << "  remove-rule <id>               Remove a rule from the rule set." << std::endl;
    stream << std::endl;
    stream << "  generate-policy                Generate a rule set (policy) based on the connected USB devices." << std::endl;
    stream << "  watch                          Watch for IPC interface events and print them to stdout." << std::endl;
    stream << "  read-descriptor                Read a USB descriptor from a file and print it in human-readable form." << std::endl;
    stream << std::endl;
    stream << "  add-user <name>                Add USBGuard IPC user/group (requires root privilges)" << std::endl;
    stream << "  remove-user <name>             Remove USBGuard IPC user/group (requires root privileges)" << std::endl;
    stream << std::endl;
  }

  static int usbguard_cli(int argc, char *argv[])
  {
    usbguard_arg0 = ::basename(argv[0]);

    if (argc == 1) {
      showTopLevelHelp();
      return EXIT_SUCCESS;
    }

    const char * subcommand_key = argv[1];
    const auto iterator = cmd_handler_map.find(subcommand_key);

    /* Not such key in the map, show list of commands */
    if (iterator == cmd_handler_map.cend()) {

      showTopLevelHelp();
      return EXIT_FAILURE;
    }


    auto subcommand = iterator->second;
    return subcommand(argc - 1, argv + 1);
  }
} /* namespace usbguard */

int main(int argc, char *argv[])
{
  try {
    return usbguard::usbguard_cli(argc, argv);
  }
  catch(const usbguard::IPCException& ex) {
    std::cerr << "IPC ERROR: request id=" << ex.messageID()
              << ": "  << ex.message() << std::endl;
  }
  catch(const usbguard::Exception& ex) {
    std::cerr << "ERROR: " << ex.message() << std::endl;
  }
  catch(const std::exception& ex) {
    std::cerr << "EXCEPTION: " << ex.what() << std::endl;
  }
  catch(...) {
    std::cerr << "BUG: Unknown exception" << std::endl;
  }
  return EXIT_FAILURE;
}

/* vim: set ts=2 sw=2 et */
